catalog/core/catalog-core-impl/metacard-type-registry/src/main/java/ddf/catalog/data/metacardtype/MetacardTypeRegistryImpl.java

```diff
@@ -110,19 +119,79 @@ private void validateInput(QualifiedMetacardType qmt) {
             String message = "QualifiedMetacardType passed in cannot be null.";
             throw new IllegalArgumentException(message);
         }
-
-        validateInput(qmt.getNamespace(), qmt.getName());
     }
 
-    private void validateInput(String namespace, String metacardTypeName) {
-        if (namespace == null) {
-            String message = "Namespace parameter cannot be null.";
-            throw new IllegalArgumentException(message);
+    /**
+     * Adds support be registered as Reference-Listener for all Services registered as MetacardType
+     * and will register/unregister those types.
+     *
+     * @param mtRef
+     */
+    public void addMetacardType(ServiceReference<MetacardType> mtRef) {
+        String id = String.valueOf((mtRef.getProperty(ID) != null ? mtRef.getProperty(ID) : ""));
+        String schema = String.valueOf((mtRef.getProperty(SCHEMA) != null ?
+                mtRef.getProperty(SCHEMA) :
+                ""));
+        MetacardType metacardType = getBundleContext().getService(mtRef);
+        boolean found = lookup(schema, id) != null || lookup(id) != null
+                || lookup(metacardType.getName()) != null;
+        if (found) {
+            LOGGER.warn(
+                    "Found new MetacardType but was already registered as, namespace: [{}]  id: [{}]",
+                    schema,
+                    id);
+        } else {
+            if (StringUtils.isNotBlank(schema) && StringUtils.isNotBlank(id)) {
+                register(new QualifiedMetacardTypeImpl(schema,
+                        id,
+                        metacardType.getAttributeDescriptors()));
+            } else if (StringUtils.isNotBlank(id)) {
+                register(new QualifiedMetacardTypeImpl("",
+                        id,
+                        metacardType.getAttributeDescriptors()));
+            } else if (StringUtils.isNotBlank(metacardType.getName())) {
+                register(new QualifiedMetacardTypeImpl(metacardType));
+            } else {
+
+            }
         }
+    }
 
-        if (metacardTypeName == null || metacardTypeName.isEmpty()) {
-            String message = "MetacardTypeName parameter cannot be null or empty.";
-            throw new IllegalArgumentException(message);
+    /**
+     * Adds support be registered as Reference-Listener for all Services registered as MetacardType
+     * and will register/unregister those types.
+     *
+     * @param mtRef
+     */
+    public void removeMetacardType(ServiceReference<MetacardType> mtRef) {
+        String id = String.valueOf((mtRef.getProperty(ID) != null ? mtRef.getProperty(ID) : ""));
```

Comment:
Might want to move this to its own method, since it's a pattern you're using in a few places.